modulemd_add_platform/modulemd_add_platform.1
.TH MODULEMD_ADD_PLATFORM "1" "" "modulemd-tools" "User Commands"
.SH NAME
modulemd_add_platform \- add a context for a new platform to modulemd-packager document
.SH SYNOPSIS
.SY modulemd_add_platform
[\fB\-\-debug\fP]
\fB\-\-old\fP \fIPLATFORM\fP
\fB\-\-new\fP \fIPLATFORM\fP
[\fB\-\-stdout\fP] [\fB\-\-skip\fP]
\fIFILE\fI
.SY modulemd_add_platform
{\fB\-h\fP|\fB\-\-help\fP}
.YS
.SH DESCRIPTION
This tool adds a new context configuration to a modulemd-packager YAML file.
The file defines how to build a module.
.P
The tool loads a \fIFILE\fP, locates all context configurations with a plaform
equivalent to the \fB\-\-old\fP argument, duplicates the configurations under
new context identifiers, and replaces there the platforms with a new one
specified in \fB\-\-new\fP argument. If a context for the new platform already
exists, the file will be kept unchanged.
.P
As the new context identifier is used the new platform value. If such context
already exists, a new, unique, small decimal number (e.g. "0", "1") will be
used.
.P
This tool aims to preserve a physical representation of the YAML files. That
includes white spaces, comments, and quoting styles. Also all fields of the
configuration, like \fBbuildrequires\fP, or \fBbuildopts\fP will be preserved.
.P
Finally, the the new content is validated and if no damage is detected, the
content is saved back to the \fIFILE\fP. Users are advised to check the
changes before commiting to them.
.SH OPTIONS
.TP
\fB\-\-old\fP \fIPLATFORM\fP
Use context configurations of this old \fIPLATFORM\fP as a template for
addition. This option is mandatory.
.TP
\fB\-\-new\fP \fIPLATFORM\fP
Set this \fIPLATFORM\fP in the new configurations. This option is mandatory.
.TP
\fB\-\-skip\fP
If specified, the files without a context configuration for the old platform and
files of modulemd-v2 format will be ignored. Otherwise, an error is reported.
.TP
\fB\-\-stdout\fP
The edited document will be printed to a standard output instead of rewriting
the \fIFILE\fP in place.
.TP
\fB\-\-debug\fP
Print various debugging messages about parsing and editing the YAML files to
an error output.
.TP
\fB\-h\fP, \fB\-\-help\fP
Show a usage text and exit without processing any files.
.SH EXAMPLE
Let's have a file \fB/tmp/test.yaml\fP:
.P
.RS
.EX
document: modulemd-packager
version: 3
data:
    configurations:
    - context: 'A'
      platform: f34
.EE
.RE
.P
Adding a configuration for \fBplatform:f35\fP can be achieved with
\fBmodulemd_add_platform \-\-old f34 \-\-new f35 /tmp/test.yaml\fP command.
The resulting file will be:
.P
.RS
.EX
document: modulemd-packager
version: 3
data:
    configurations:
    - context: 'A'
      platform: f34
    - context: 'f35'
      platform: f35
.EE
.RE
.SH EXIT CODE
Return a zero exit code if the \fIFILE\fP was successfully edited or the
\fIFILE\fP already contained the new platform.
.P
Return non-zero exit code if an error occurred.
.P
If the input \fBFILE\fP does not contain a context for the old platform, or if
the file is modulemd-v2 document, which does not use context configurations,
the behaviour depends on a use of \fB\-\-skip\fP option. If it isn't used,
a non-zero code will be returned. Otherwise, zero code will be returned.
.SH REPORTING BUGS
Bugs and feature requests can be reported at the
.UR https://github.com/rpm-software-management/modulemd-tools
modulemd-tools home page
.UE .
.SH SEE ALSO
Specification of the modulemd-packager documents is available among
a documentation for \fBlibmodulemd\fP package or in \fByaml_spec\fP directory
at the
.UR https://github.com/fedora-modularity/libmodulemd
libmodulemd home page
.UE .


modulemd_add_platform/modulemd_add_platform/modulemd_add_platform.py
#!/usr/bin/python3
import sys
import argparse
import os
from pathlib import Path
import re
import stat
import tempfile
import gi
gi.require_version('Modulemd', '2.0')
from gi.repository import Modulemd
#from gi.repository import Glib

def dequote_yaml_string(input):
    """Remove string separators from a YAML string scalar.

    E.g. input is "'FOO'", output is "FOO". In case of error returns None.
    """
    # TODO: backslash escaping
    result = re.match(r'^\s*"([^"]*)"', input)
    if result:
        return result.group(1)
    result = re.match(r'^\s*\'([^\']*)\'', input)
    if result:
        return result.group(1)
    result = re.match(r'^\s*([^"\'#]+)', input)
    if result:
        return result.group(1)

def edit(content, old_platform, new_platform, context_map):
    """Manually add build configurations in YAML document.

    Content is a modulemd-packager-v3 YAML document with contexts of for
    old_plaform. This function duplicates contexts listed in context_map keys
    to new contexts listed in context_map values. When duplicating, it will
    rewrite old_platform to new_platform.
    """
    # TODO: There can be multiple contexts for a platform. All of them should
    # be duplicated.
    output = []
    record = []
    contexts = []
    in_configurations = False
    in_context = False
    current_context = ''
    this_context_is_old_platform = False
    new_context_starts_at_line_number = -1;
    old_context_lines = []
    for line in content.splitlines():
        print(line)
        output.append(line)

        # Comments can interleave disrespecting indentation
        if re.match(r'^\s*#', line):
            if in_context:
                record.append(line)
            continue

        if in_context:
            result = re.match(
                    r'^' + indent_configurations + indent_context + r'\S',
                    line)
            if result:
                result = re.match(
                        r'^(\s+platform\s*:\s*)([\'"]?)' + old_platform +
                            r'(\2)(\s.*|#.*|$)',
                        line)
                if result:
                    print('HIT old platform')
                    this_context_is_old_platform = True
                    line = result.group(1) + result.group(2) + new_platform \
                            + result.group(2)
                record.append(line)
                continue
            else:
                in_context = False
                print('END context of ' + current_context)
                for x in record:
                    print('RECORDED: ' + x)
                if current_context in context_map:
                    # Insert the recorded context in before the last output line
                    for x in record:
                        output.insert(-1, x)

        if in_configurations:
            # TODO: Handle "-\n context". Comments can interleave.
            result = re.match(
                    r'^(' + indent_configurations +
                        r'(\s*)-(\s+)context\s*:\s*)(\S+)',
                    line)
            if result:
                in_context = True
                context_value_prefix = result.group(1)
                indent_context = result.group(2) + ' ' + result.group(3)
                current_context = dequote_yaml_string(result.group(4))
                contexts.append(current_context)
                record.clear()
                if current_context in context_map:
                    record.append(context_value_prefix + "'"
                            + context_map[current_context] +"'")
                print('START context "{}"'.format(current_context))
            continue

        # TODO: Restrict the space prefix to a second level
        result = re.match(r'^(\s+)configurations\s*:', line)
        if result:
            in_configurations = True
            indent_configurations = result.group(1)
            print('START configurations')
            continue

    print('OUTPUT:')
    for line in output:
        print(line)

    return '\n'.join(output)

def equaled_modulemd_packager(a, b):
    """Compare two modulemd-packager documents.

    Return true if equaled. False otherwise. It modifies the documents.
    """
    # convert_to_index() requires mandatory fields (e.g. summary) to be set.
    # It seems that validation in packager_read_*() has a bug.
    for document in a, b:
        document.set_summary('dummy')
        document.set_description('dummy')
    return a.convert_to_index().dump_to_string() \
            == b.convert_to_index().dump_to_string()

def generate_context(contexts):
    """Generate a new string which does not exist in the given list."""
    for integer in range():
        context = '{!s}'.format(integer)
        if context not in contexts:
            return context

def duplicate_configuration(template_configuration, new_context, new_platform):
    """Copy a template configuration, set the new context, and return it."""
    new_configuration = template_configuration.copy()
    new_configuration.set_context(new_context)
    new_configuration.set_platform(new_platform)
    return new_configuration

def process_string(content, old_platform, new_platform):
    """Add a configuration for a new platform to the modulemd document string.

    It returns an error code and a string.
    In case of no error, code will be 0 and the string will be the processed,
    output document.
    In case of an error, code will nonzero and the string will contain an
    error message.
    A special error -1 means the document needs no edition.
    """

    # Parse and validate the content
    try:
        document = Modulemd.read_packager_string(content)
    except Exception as e:
        return 1, 'Unable to parse a modulemd-packager document: {}'.format(e)

    # Enumerate all contexts with the old platform
    contexts = document.get_build_config_contexts_as_strv()
    template_configurations = []
    for context in contexts:
        configuration = document.get_build_config(context)
        if configuration.get_platform() == new_platform:
            # If there is already a context for the new platform,
            # there no work for us.
            return -1, 'A context for the new {} platform already exists: {}' \
                .format(configuration.get_context(), new_platform)
        if configuration.get_platform() == old_platform:
            template_configurations.append(configuration)

    # Duplicate configurations
    new_configurations = []
    old_to_new_context_map = {}
    if len(template_configurations) == 0:
        return 2, 'No context with the old platform {}.'.format(old_platform)
    elif len(template_configurations) == 1:
        # Try using the new platform as the new context
        if new_platform in contexts:
            new_context = generate_context(contexts)
        else:
            new_context = new_platform
        new_configurations.append(duplicate_configuration(
                template_configurations[0], new_context, new_platform))
        contexts.append(new_context)
        old_to_new_context_map[template_configurations[0].get_context()] = \
                new_context
    else:
        for template_configuration in template_configurations:
            new_context = generate_context(contexts)
            new_configurations.append(duplicate_configuration(
                    template_configuration, new_context, new_platform))
            contexts.append(new_context)
            old_to_new_context_map[template_configuration.get_context()] = \
                    new_context
    for new_configuration in new_configurations:
        document.add_build_config(new_configuration)

    # Edit the document preserving formatting
    edited_content = edit(content, old_platform, new_platform,
            old_to_new_context_map)

    # Reparse the document to verify it was not damaged
    try:
        edited_document = Modulemd.read_packager_string(edited_content)
    except Exception as e:
        return 3, 'Unable to parse the edited document: {}'.format(e)

    # Compare library-edited and manually edited documents
    if not equaled_modulemd_packager(document, edited_document):
        return 4, 'Editing would demage the modulemd-packager document.'

    return 0, edited_content

def process_file(file, old_platform, new_platform):
    """Add a configuration for a new platform to the modulemd file.

    In case of error, return the error message.
    In case of success, return (None, notice), there warning is an optinal
    notification the use could be interrested in.
    """

    # Open the modulemd-packager file
    try:
        fd = open(file, encoding='UTF-8')
    except Exception as e:
        return 'Could not open the modulemd-packager file: {}'.format(e)
    # Read the file
    try:
        content = fd.read()
    except Exception as e:
        fd.close()
        return '{}: Could not read the modulemd-packager file: {}'.format(file,
                e)
    # Retrieve permissions of the file
    try:
        stat = os.fstat(fd.fileno())
    except:
        fd.close()
        return '{}: Could not stat the modulemd-packager file: {}'.format(file, e)
    # Close the file
    fd.close()

    # Edit the document in memory
    error, text = process_string(content, old_platform, new_platform)
    if error == -1:
        return None, '{}: Skipped: {}'.format(file, text)
    elif error:
        return '{}: {}'.format(file, text)
    # TODO: Handle soft errors by printing/keeping the edited text (damage etc.)

    # Store the edited document into a temporary file
    try:
        temp_fd, temp_name = tempfile.mkstemp(dir=Path(file).parent, text=True)
        temp_file = os.fdopen(temp_fd, mode='w', encoding='UTF-8')
        temp_file.write(text)
    except Exception as e:
        temp_file.close()
        return '{}: Could not write to a temporary file: {}'.format(temp_name,
                e)
    # Copy file permissions
    try:
        os.fchmod(temp_fd, stat.st_mode)
    except Exception as e:
        temp_file.close()
        return '{}: Could not copy a file mode: {}'.format(temp_name, e)
    try:
        os.fchown(temp_fd, stat.st_uid, stat.st_gid)
    except Exception as e:
        temp_file.close()
        return '{}: Could not copy a file ownership: {}'.format(temp_name, e)
    # Close the descriptor
    try:
        temp_file.close()
    except Exception as e:
        return '{}: Could not close the file: {}'.format(temp_name, e)
    # And replace the file
    try:
        os.replace(temp_name, file)
    except Exception as e:
        return '{}: Could not rename to {}: {}'.format(temp_name, file, e)

def main():
    arg_parser = argparse.ArgumentParser(
        description = 'Add a context for the given platform')
    arg_parser.add_argument('file')
    arg_parser.add_argument('--old', required=True, metavar='PLATFORM',
            help='old platform')
    arg_parser.add_argument('--new', required=True, metavar='PLATFORM',
            help='new platform')
    arguments = arg_parser.parse_args()
    # TODO: Validate platform values
    error, warning = process_file(arguments.file, arguments.old, arguments.new)
    if error:
        sys.stderr.write('Error: {}\n'.format(error))
        exit(1)
    if warning:
        sys.stderr.write('{}\n'.format(warning))


if __name__ == '__main__':
    main()
